pkg/istio-agent/healthcheck.go

```diff
@@ -0,0 +1,227 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istioagent
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"time"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/genproto/googleapis/rpc/status"
+)
+
+type HealthCheckType string
+
+const (
+	HTTPHealthCheck   HealthCheckType = "HTTP"
+	TCPHealthCheck    HealthCheckType = "TCP"
+	ExecHealthCheck   HealthCheckType = "Exec"
+	HealthInfoTypeURL string          = "type.googleapis.com/istio.v1.HealthInformation"
+)
+
+type WorkloadHealthChecker struct {
+	config ApplicationHealthCheckConfig
+}
+
+type ApplicationHealthCheckConfig struct {
+	InitialDelay   time.Duration
+	ProbeTimeout   time.Duration
+	CheckFrequency time.Duration
+	SuccessThresh  int
+	FailThresh     int
+	CheckType      HealthCheckType
+	HTTPConfig     HTTPHealthCheckConfig
+	TCPConfig      TCPHealthCheckConfig
+	ExecConfig     ExecHealthCheckConfig
+}
+
+type HTTPHealthCheckConfig struct {
+	Path    string
+	Port    uint32
+	Scheme  string
+	Headers map[string][]string
```

Comment:
I think key-value is enough

Reply:
Maybe can add host field to process `Host` explicitly

Reply:
make sense

Reply:
Can we use the same type here?

Reply:
this is just for quick iteration @ramaraochavali. the actual fields (introduced in https://github.com/istio/api/pull/1620). this will be removed soon, its just to test essentially.

Reply:
but i will change it, sure